java/org/apache/tomcat/util/net/SecureNio2Channel.java

```diff
@@ -70,6 +73,8 @@
     protected boolean closed;
     protected boolean closing;
 
+    private Map<String,List<String>> additionalTlsAttributes = new HashMap<>();
```

Comment:
could be `final`

Reply:
I see there are some `volatile` member fields in this class.  Should `sniComplete` be `volatile` as well ?

Reply:
Is there a reason to use a Map of String-Lists here, instead of having a more concrete object with well-defined members? Hash maps are quick, but since we know the "names" of these things in advance, why not use a custom class for this purpose?

Reply:
Yes, sniComplete should be volatile.

The reason to use Map<String,List<String>> is that the attribute is exposed to user code so I don;t want it to be an internal Tomcat class. If we get this added to the spec then we could define a class for it in the spec.

Reply:
👍